### PR TITLE
Generate processes with event-based gateways

### PR DESCRIPTION
**Pull request by saig0**
## Description

* extend the block builder by event-based gateways
* for now, event-based gateways will have at least one connected message catch event
  * no need to add other event types to verify that the event-based gateway works (in general)
  * we should add timer events later   

process-with-event-based-gateway

## Related issues

Prepare #6191

## Definition of Done

_Not all items need to be done depending on the issue and the pull request._

Code changes:
* [ ] The changes are backwards compatibility with previous versions
* [ ] If it fixes a bug then PRs are created to [backport](https://github.com/zeebe-io/zeebe/compare/stable/0.24...develop?expand=1&template=backport_template.md&title=[Backport%200.24]) the fix to the last two minor versions. You can trigger a backport by assigning labels (e.g. `backport stable/0.25`) to the PR, in case that fails you need to create backports manually.

Testing:
* [ ] There are unit/integration tests that verify all acceptance criterias of the issue
* [ ] New tests are written to ensure backwards compatibility with further versions
* [ ] The behavior is tested manually
* [ ] The change has been verified by a QA run
* [ ] The impact of the changes is verified by a benchmark 

Documentation: 
* [ ] The documentation is updated (e.g. BPMN reference, configuration, examples, get-started guides, etc.)
* [ ] New content is added to the [release announcement](https://drive.google.com/drive/u/0/folders/1DTIeswnEEq-NggJ25rm2BsDjcCQpDape)

### CODE REVIEW

**Comment by saig0** [comment]
The `ReplayStateRandomizedPropertyTest` fails. It seems that we have already problems with the replay if an event-based gateway is triggered :sweat_smile:

**Comment by Zelldon** [comment]
What a surprise @saig0 :laughing:

**Comment by saig0** [comment]
@korthout please have a look at the changes. I'll merge the PR after #6191 is done.

File: test-util/src/main/java/io/zeebe/test/util/bpmn/random/blocks/EventBasedGatewayBlockBuilder.java [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+
+    final var blockSequenceBuilderFactory = context.getBlockSequenceBuilderFactory();
+
+    // must have at least two branches
+    final int numberOfBranches = Math.max(2, random.nextInt(maxBranches));
+
+    for (int i = 0; i < numberOfBranches; i++) {
+      final var branchId = idGenerator.nextId();
+      final var branchBuilder =
+          blockSequenceBuilderFactory.createBlockSequenceBuilder(context.withIncrementedDepth());
```
Comment (korthout) [inline test-util/src/main/java/io/zeebe/test/util/bpmn/random/blocks/EventBasedGatewayBlockBuilder.java:48]: Are you sure the depth should be incremented here? If so I think you should also check the currentDepth and the maxDepth and whether you're allowed to go deeper. (See SubProcessBlockBuilder for example). Personally, I think this does not increase the depth. In that case also remember to disable depth in the Factory.

**Review by korthout — APPROVED** [review]
@saig0 LGTM just 1 question regarding depth. Please have a look.

File: test-util/src/main/java/io/zeebe/test/util/bpmn/random/blocks/EventBasedGatewayBlockBuilder.java (hunk 1 above)
Comment (saig0) [reply test-util/src/main/java/io/zeebe/test/util/bpmn/random/blocks/EventBasedGatewayBlockBuilder.java:48]: I'm not sure :sweat_smile: I looked at the parallel and exclusive gateway and did the same here. 

@pihme does a gateway adding depth? and why?

Comment (pihme) [reply test-util/src/main/java/io/zeebe/test/util/bpmn/random/blocks/EventBasedGatewayBlockBuilder.java:48]: It does add depth - in a way. It's more a technical reason than a model reason.

Think of depth here as layers in a call stack and not nesting in the BPMN model.

If the gateway wouldn't add depth, you could generate infinite processes: A process which contains a gateway which on its default branch contains a gateway, which on its default branch contains a gateway ...

Maybe this issue is now less pronounced - now that we have more elements which can be added to a process. But when I fist wrote it and there were only a couple of elements, the processes exploded in complexity.

**Pull request merged**